File: app/src/main/java/com/electricdreams/shellshock/feature/history/PaymentsHistoryActivity.java
package com.electricdreams.shellshock.feature.history;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.PopupMenu;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.electricdreams.shellshock.R;
import com.electricdreams.shellshock.core.data.model.PaymentHistoryEntry;
import com.electricdreams.shellshock.ui.adapter.PaymentsHistoryAdapter;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

public class PaymentsHistoryActivity extends AppCompatActivity {
    private static final String PREFS_NAME = "PaymentHistory";
    private static final String KEY_HISTORY = "history";

    private PaymentsHistoryAdapter adapter;
    private TextView emptyView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_history);

        // Setup Back Button
        View backButton = findViewById(R.id.back_button);
        if (backButton != null) {
            backButton.setOnClickListener(v -> finish());
        }

        // Setup RecyclerView
        RecyclerView recyclerView = findViewById(R.id.history_recycler_view);
        emptyView = findViewById(R.id.empty_view);
        
        adapter = new PaymentsHistoryAdapter();
        adapter.setOnItemClickListener(this::showTransactionDetails);
        
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        // Load and display history
        loadHistory();
    }

    private void showTransactionDetails(PaymentHistoryEntry entry, int position) {
        BottomSheetDialog sheet = new BottomSheetDialog(this);
        sheet.setContentView(R.layout.dialog_transaction_details);

        TextView amountText = sheet.findViewById(R.id.detail_amount);
        TextView dateText = sheet.findViewById(R.id.detail_date);
        TextView tokenText = sheet.findViewById(R.id.detail_token);
        Button copyButton = sheet.findViewById(R.id.btn_copy);
        Button openWithButton = sheet.findViewById(R.id.btn_open_with);
        Button deleteButton = sheet.findViewById(R.id.btn_delete);

        if (amountText != null) {
            long amount = entry.getAmount();
            String sign = amount >= 0 ? "+" : "";
            amountText.setText(String.format(Locale.getDefault(), "%s $%d", sign, Math.abs(amount)));
        }
        
        if (dateText != null) {
            SimpleDateFormat dateFormat = new SimpleDateFormat("MMM d, yyyy HH:mm", Locale.getDefault());
            dateText.setText(dateFormat.format(entry.getDate()));
        }

        if (tokenText != null) {
            tokenText.setText(entry.getToken());
        }

        if (copyButton != null) {
            copyButton.setOnClickListener(v -> {
                ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
                ClipData clip = ClipData.newPlainText("Payment", entry.getToken());
                clipboard.setPrimaryClip(clip);
                Toast.makeText(this, "Payment copied to clipboard", Toast.LENGTH_SHORT).show();
                sheet.dismiss();
            });
        }

        if (openWithButton != null) {
            openWithButton.setOnClickListener(v -> {
                openPaymentWithApp(entry.getToken());
                sheet.dismiss();
            });
        }

        if (deleteButton != null) {
            deleteButton.setOnClickListener(v -> {
                sheet.dismiss();
                showDeleteConfirmation(entry, position);
            });
        }

        sheet.show();
    }

    private void openPaymentWithApp(String token) {
        String cashuUri = "cashu:" + token;
        Intent uriIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(cashuUri));
        uriIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

        Intent shareIntent = new Intent(Intent.ACTION_SEND);
        shareIntent.setType("text/plain");
        shareIntent.putExtra(Intent.EXTRA_TEXT, cashuUri);
        
        Intent chooserIntent = Intent.createChooser(uriIntent, "Open payment with...");
        chooserIntent.putExtra(Intent.EXTRA_INITIAL_INTENTS, new Intent[] { shareIntent });
        
        try {
            startActivity(chooserIntent);
        } catch (Exception e) {
            Toast.makeText(this, "No apps available to handle this payment", Toast.LENGTH_SHORT).show();
        }
    }

    private void showDeleteConfirmation(PaymentHistoryEntry entry, int position) {
        new AlertDialog.Builder(this)
            .setTitle("Delete Payment")
            .setMessage("Are you sure you want to delete this payment from history?")
            .setPositiveButton("Delete", (dialog, which) -> deletePaymentFromHistory(position))
            .setNegativeButton("Cancel", null)
            .show();
    }

    private void showClearHistoryConfirmation() {
        new AlertDialog.Builder(this)
            .setTitle("Clear History")
            .setMessage("Are you sure you want to clear all payment history? This action cannot be undone.")
            .setPositiveButton("Clear All", (dialog, which) -> clearAllHistory())
            .setNegativeButton("Cancel", null)
            .show();
    }

    private void loadHistory() {
        List<PaymentHistoryEntry> history = getPaymentHistory();
        Collections.reverse(history); // Show newest first
        adapter.setEntries(history);
        
        boolean isEmpty = history.isEmpty();
        if (emptyView != null) {
            emptyView.setVisibility(isEmpty ? View.VISIBLE : View.GONE);
        }
    }

    private void clearAllHistory() {
        SharedPreferences prefs = getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putString(KEY_HISTORY, "[]");
        editor.apply();
        loadHistory();
    }

    private void deletePaymentFromHistory(int position) {
        List<PaymentHistoryEntry> history = getPaymentHistory();
        Collections.reverse(history);
        if (position >= 0 && position < history.size()) {
            history.remove(position);
            Collections.reverse(history);
            
            SharedPreferences prefs = getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
            SharedPreferences.Editor editor = prefs.edit();
            editor.putString(KEY_HISTORY, new Gson().toJson(history));
            editor.apply();
            
            loadHistory();
        }
    }

    public static List<PaymentHistoryEntry> getPaymentHistory(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
        String json = prefs.getString(KEY_HISTORY, "[]");
        Type type = new TypeToken<ArrayList<PaymentHistoryEntry>>(){}.getType();
        return new Gson().fromJson(json, type);
    }

    private List<PaymentHistoryEntry> getPaymentHistory() {
        return getPaymentHistory(this);
    }

    public static void addToHistory(Context context, String token, long amount) {
        List<PaymentHistoryEntry> history = getPaymentHistory(context);
        history.add(new PaymentHistoryEntry(token, amount, new java.util.Date()));
        
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putString(KEY_HISTORY, new Gson().toJson(history));
        editor.apply();
    }
}

File: app/src/main/java/com/electricdreams/shellshock/feature/history/PaymentsHistoryActivity.kt
package com.electricdreams.shellshock.feature.history

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.widget.Toast
import androidx.activity.compose.setContent
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.compose.runtime.mutableStateOf
import com.electricdreams.shellshock.core.data.model.PaymentHistoryEntry
import com.electricdreams.shellshock.ui.screens.HistoryScreen
import com.electricdreams.shellshock.ui.theme.CashAppTheme
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.util.Collections
import java.util.Date

class PaymentsHistoryActivity : AppCompatActivity() {

    private val historyState = mutableStateOf<List<PaymentHistoryEntry>>(emptyList())
    private var isNightMode = false

    override fun onCreate(savedInstanceState: Bundle?) {
        // Load theme preference
        val prefs = getSharedPreferences(PREFS_NAME, MODE_PRIVATE)
        isNightMode = prefs.getBoolean(KEY_NIGHT_MODE, false)
        AppCompatDelegate.setDefaultNightMode(if (isNightMode) AppCompatDelegate.MODE_NIGHT_YES else AppCompatDelegate.MODE_NIGHT_NO)

        super.onCreate(savedInstanceState)

        setContent {
            CashAppTheme(darkTheme = isNightMode) {
                HistoryScreen(
                    history = historyState.value,
                    onBackClick = { finish() },
                    onClearHistoryClick = { showClearHistoryConfirmation() },
                    onCopyClick = { token -> copyToClipboard(token) },
                    onOpenClick = { token -> openPaymentWithApp(token) },
                    onDeleteClick = { entry -> showDeleteConfirmation(entry) }
                )
            }
        }

        loadHistory()
    }

    private fun loadHistory() {
        val history = getPaymentHistory(this)
        history.reverse() // Show newest first
        historyState.value = history
    }

    private fun copyToClipboard(token: String) {
        val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val clip = ClipData.newPlainText("Payment", token)
        clipboard.setPrimaryClip(clip)
        Toast.makeText(this, "Payment copied to clipboard", Toast.LENGTH_SHORT).show()
    }

    private fun openPaymentWithApp(token: String) {
        val cashuUri = "cashu:$token"
        
        // Create intent for viewing the URI
        val uriIntent = Intent(Intent.ACTION_VIEW, Uri.parse(cashuUri))
        uriIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

        // Create a fallback intent for sharing as text
        val shareIntent = Intent(Intent.ACTION_SEND)
        shareIntent.type = "text/plain"
        shareIntent.putExtra(Intent.EXTRA_TEXT, cashuUri)
        
        // Combine both intents into a chooser
        val chooserIntent = Intent.createChooser(uriIntent, "Open payment with...")
        chooserIntent.putExtra(Intent.EXTRA_INITIAL_INTENTS, arrayOf(shareIntent))
        
        try {
            startActivity(chooserIntent)
        } catch (e: Exception) {
            Toast.makeText(this, "No apps available to handle this payment", Toast.LENGTH_SHORT).show()
        }
    }

    private fun showDeleteConfirmation(entry: PaymentHistoryEntry) {
        AlertDialog.Builder(this)
            .setTitle("Delete Payment")
            .setMessage("Are you sure you want to delete this payment from history?")
            .setPositiveButton("Delete") { _, _ -> deletePayment(entry) }
            .setNegativeButton("Cancel", null)
            .show()
    }

    private fun deletePayment(entryToDelete: PaymentHistoryEntry) {
        val history = getPaymentHistory(this)
        // Find and remove the entry (comparing by token and date ideally, but object reference might fail if reloaded)
        // Let's find by token and amount for now
        val iterator = history.iterator()
        while (iterator.hasNext()) {
            val item = iterator.next()
            if (item.token == entryToDelete.token && item.amount == entryToDelete.amount) {
                iterator.remove()
                break
            }
        }
        
        saveHistory(history)
        loadHistory()
    }

    private fun showClearHistoryConfirmation() {
        AlertDialog.Builder(this)
            .setTitle("Clear History")
            .setMessage("Are you sure you want to clear all payment history? This action cannot be undone.")
            .setPositiveButton("Clear All") { _, _ -> clearAllHistory() }
            .setNegativeButton("Cancel", null)
            .show()
    }

    private fun clearAllHistory() {
        saveHistory(ArrayList())
        loadHistory()
    }

    private fun saveHistory(history: List<PaymentHistoryEntry>) {
        val prefs = getSharedPreferences(PREFS_NAME, MODE_PRIVATE)
        val editor = prefs.edit()
        editor.putString(KEY_HISTORY, Gson().toJson(history))
        editor.apply()
    }

    companion object {
        private const val PREFS_NAME = "PaymentHistory"
        private const val KEY_HISTORY = "history"
        private const val KEY_NIGHT_MODE = "night_mode"

        fun getPaymentHistory(context: Context): ArrayList<PaymentHistoryEntry> {
            val prefs = context.getSharedPreferences(PREFS_NAME, MODE_PRIVATE)
            val json = prefs.getString(KEY_HISTORY, "[]")
            val type = object : TypeToken<ArrayList<PaymentHistoryEntry>>() {}.type
            return Gson().fromJson(json, type)
        }

        fun addToHistory(context: Context, token: String, amount: Long) {
            val history = getPaymentHistory(context)
            history.add(PaymentHistoryEntry(token, amount, Date()))
            
            val prefs = context.getSharedPreferences(PREFS_NAME, MODE_PRIVATE)
            val editor = prefs.edit()
            editor.putString(KEY_HISTORY, Gson().toJson(history))
            editor.apply()
        }
    }
}
